Make FDWatcher check for file descriptors equal to -1

JamesWrigley opened the pull request:
This is almost definitely an invalid file descriptor. And if it's not then that means the process is running out of file descriptors anyway.

I ran into this when my code passed `poll_fd()` a `RawFD(-1)` from a C library, which caused a rather confusing error:
```julia
1-element ExceptionStack:                                                                                                                                                                                                                      
BoundsError: attempt to access 0-element Vector{Any} at index [0]                                                                                                                                                                              
Stacktrace:                                                                                                                                                                                                                                    
 [1] getindex                                                                                                                                                                                                                                  
   @ Base ./essentials.jl:13 [inlined]                                                                                                                                                                                                         
 [2] FileWatching._FDWatcher(fd::RawFD, readable::Bool, writable::Bool)                                                                                                                                                                        
   @ FileWatching /cache/build/builder-amdci4-6/julialang/julia-release-1-dot-10/usr/share/julia/stdlib/v1.10/FileWatching/src/FileWatching.jl:176                                                                                             
 [3] FileWatching._FDWatcher(fd::RawFD, readable::Bool, writable::Bool)                                                                                                                                                                        
   @ FileWatching /cache/build/builder-amdci4-6/julialang/julia-release-1-dot-10/usr/share/julia/stdlib/v1.10/FileWatching/src/FileWatching.jl:165 [inlined]                                                                                   
 [4] poll_fd(s::RawFD, timeout_s::Int64; readable::Bool, writable::Bool)                                                                                                                                                                       
   @ FileWatching /cache/build/builder-amdci4-6/julialang/julia-release-1-dot-10/usr/share/julia/stdlib/v1.10/FileWatching/src/FileWatching.jl:709                                                                                             
 [5] poll_fd (repeats 2 times)                                                                                                                                                                                                                 
   @ /cache/build/builder-amdci4-6/julialang/julia-release-1-dot-10/usr/share/julia/stdlib/v1.10/FileWatching/src/FileWatching.jl:706 [inlined]
```

JamesWrigley committed:
Make FDWatcher check for file descriptors equal to -1

This is almost definitely an invalid file descriptor. And if it's not then that
means the process is running out of file descriptors anyway.

JamesWrigley committed:
fixup! Make FDWatcher check for file descriptors equal to -1

JamesWrigley commented:
Lots of failures :smiling_face_with_tear: AFAICT the only one related to this PR is the windows failures, restricted the test to non-windows in 0a1cfe724d (on windows it throws another exception immediately anyway).

JamesWrigley commented:
A couple of failures seem to be because of artifacts not being downloaded? And one ran out of memory. But it looks like the FileWatching tests all passed.

vtjnash commented on stdlib/FileWatching/src/FileWatching.jl:
```diff
@@ -164,9 +164,12 @@ mutable struct _FDWatcher
             _FDWatcher(fd::RawFD, mask::FDEvent) = _FDWatcher(fd, mask.readable, mask.writable)
             function _FDWatcher(fd::RawFD, readable::Bool, writable::Bool)
-                if !readable && !writable
+                if fd == RawFD(-1)
+                    throw(ArgumentError("Passed file descriptor is $(fd) == RawFD(-1), this is probably not a valid file descriptor"))
+                elseif !readable && !writable
                     throw(ArgumentError("must specify at least one of readable or writable to create a FDWatcher"))
                 end
+
                 fdnum = Core.Intrinsics.bitcast(Int32, fd) + 1
```
```suggestion
                fdnum = Core.Intrinsics.bitcast(Int32, fd) + 1
                if fdnum <= 0
                    throw(ArgumentError("file descriptor fd=$fd is not a valid file descriptor"))
                elseif !readable && !writable
                    throw(ArgumentError("must specify at least one of readable or writable to create a FDWatcher"))
                end

```

JamesWrigley replied:
Fixed in ed06cc419c.

vtjnash commented on stdlib/FileWatching/src/FileWatching.jl:
```diff
@@ -232,6 +235,10 @@ mutable struct _FDWatcher
     @static if Sys.iswindows()
         _FDWatcher(fd::RawFD, mask::FDEvent) = _FDWatcher(fd, mask.readable, mask.writable)
         function _FDWatcher(fd::RawFD, readable::Bool, writable::Bool)
+            if fd == RawFD(-1)
+                throw(ArgumentError("Passed file descriptor is $(fd) == RawFD(-1), this is probably not a valid file descriptor"))
+            end
```
```suggestion
            fdnum = Core.Intrinsics.bitcast(Int32, fd) + 1
            if fdnum <= 0
                throw(ArgumentError("file descriptor fd=$fd is not a valid file descriptor"))
            end
```

JamesWrigley replied:
Fixed in ed06cc419c, but why do we add 1 to the value? If I understand the previous constructor correctly it's because we use `fdnum` as an index, but that's not done here.

vtjnash replied:
Just to avoid a spurious difference with the unix code

vtjnash commented on stdlib/FileWatching/src/FileWatching.jl:
```diff
@@ -232,6 +235,10 @@ mutable struct _FDWatcher
+
             handle = Libc._get_osfhandle(fd)
             return _FDWatcher(handle, readable, writable)
         end
```
And github won't let me comment on the following line, but adding the same to the next function too:
```julia
        function _FDWatcher(fd::WindowsRawSocket, readable::Bool, writable::Bool)
            if fd == Base.INVALID_OS_HANDLE
                throw(ArgumentError("file descriptor fd=$fd is not a valid file descriptor"))
            elseif !readable && !writable
```

JamesWrigley replied:
Added in ed06cc419c.

vtjnash reviewed (commented):
Sorry this slipped through the cracks while I was on vacation. It seems good to me. I think it might be better to write `fd=RawFD(4294967295)`  instead of `Base.RawFD(4294967295) == RawFD(-1)` though? I made some inline suggestions, but what do you think of those?

But secondly, change the show method:
```
libc.jl:44  show(io::IO, fd::RawFD) = print(io, "RawFD(", bitcast(UInt32, fd), ')')  # avoids invalidation via show_default
```
To bitcast to Int32 instead of UInt32, so that it prints as a signed number

JamesWrigley committed:
Print RawFD's as signed integers

Sometimes we get things like RawFD(-1) and it's a bit confusing for them to be
printed as a giant UInt32.

JamesWrigley committed:
fixup! Make FDWatcher check for file descriptors equal to -1

JamesWrigley commented:
> Sorry this slipped through the cracks while I was on vacation. It seems good to me. I think it might be better to write `fd=RawFD(4294967295)` instead of `Base.RawFD(4294967295) == RawFD(-1)` though? I made some inline suggestions, but what do you think of those?

No worries, thanks for the review :) And yeah that makes sense, I changed the message in ed06cc419c.

> But secondly, change the show method:
> 
> ```
> libc.jl:44  show(io::IO, fd::RawFD) = print(io, "RawFD(", bitcast(UInt32, fd), ')')  # avoids invalidation via show_default
> ```
> 
> To bitcast to Int32 instead of UInt32, so that it prints as a signed number

Fixed in 4374145a64.
